File: p200.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void fill(vector<vector<char>>& grid, int row, int col) {
        bool out_of_bounds = row < 0 || row >= grid.size() || col < 0 || col >= grid[0].size();
        if(out_of_bounds || grid[row][col] == '0') {
            return;
        }
        grid[row][col] = '0';
        fill(grid, row+1, col);
        fill(grid, row-1, col);
        fill(grid, row, col+1);
        fill(grid, row, col-1);
    }
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        for(int i = 0; i < grid.size(); ++i) {
            for(int j = 0; j < grid[0].size(); ++j) {
                if(grid[i][j] == '1') {
                    ++count;
                    fill(grid, i, j);
                }
            }
        }
        return count;
    }
};

File: p200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0]) if len(grid) > 0 else 0

        if m == 0 or n == 0:
            return 0

        def fill(i, j):
            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == '0':
                return
            grid[i][j] = '0'
            fill(i, j+1)
            fill(i, j-1)
            fill(i+1, j)
            fill(i-1, j)

        ans = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    ans += 1
                    fill(i, j)

        return ans
